doc/reference/task_wdt/index.rst

```diff
@@ -0,0 +1,53 @@
+.. _task_wdt:
+
+Task Watchdog
+#############
+
+Overview
+********
+
+Many microcontrollers feature a hardware watchdog timer peripheral. Its purpose
+is to trigger an action (usually a system reset) in case of severe software
+malfunctions. Once initialized, the watchdog timer has to be restarted ("fed")
+in regular intervals to prevent it from timing out. If the software got stuck
+and does not manage to feed the watchdog anymore, the corrective action is
+triggered to bring the system back to normal operation.
+
+In real-time operating systems with multiple tasks running in parallel, a
+single watchdog instance may not be sufficient anymore, as it can be used for
+only one task. This software watchdog based on kernel timers provides a method
+to supervise multiple threads or tasks (called watchdog channels).
+
+An existing hardware watchdog can be used as an optional fallback if the task
+watchdog itself or the scheduler has a malfunction.
+
+The task watchdog uses a kernel timer as its backend. If configured properly,
+the timer ISR is never actually called during normal operation, as the timer is
+continuously updated in the feed calls.
+
+It's currently not possible to have multiple instances of task watchdogs.
+Instead, the task watchdog API can be accessed globally to add or delete new
+channels without passing around a context or device pointer in the firmware.
```

Comment:
Is this a hard requirement on the API? It seems like an odd limitation compared to many other Zephyr components.

Take for instance a use-case of having two hardware watchdogs; one for resetting the MCU itself and one for keeping an external device out of reset (e.g. an NXP Kinetis External Watchdog Monitor (EWM)). The latter is useful in safety focused applications such as motor control. It would be nice to be able to support both of these watchdogs with a "multiplexer" like the task_wdt.

Reply:
I can appreciate the desire, but I think this API's scope is all about the CPU reset.  And remember that its intention is to function with or without an actual hardware CPU watchdog.  Supporting a job-specific watchdog just doesn't really seem to be the intent of this API.

Reply:
I consider the task watchdog as something similar to a kernel service. Like the kernel timer, it hides the used hardware (system clock) in the background and has a global context, so that a new channel can be added anywhere in the application. In contrast to actual devices, we can't use `device_get_binding` in each thread to get a specific instance, that's why I prefer the global scope.

For your use case I'm wondering if it would not be sufficient to use the existing watchdog API? I guess that the EWM should mainly supervise the control loop and could still be used in parallel.

Also other RTOSes targeting safety-critical applications ([VxWorks](https://www.ee.ryerson.ca/~courses/ee8205/Data-Sheets/Tornado-VxWorks/vxworks/ref/wdLib.html), [Erika Enterprise](http://erika.tuxfamily.org/wiki/index.php?title=AUTOSAR-like_WDG_Driver)) seem to support only a single (global) task watchdog. This API does already provide more features than most other implementations, so I'd rather try to not make it too complex.